### DISCUSSION TITLE
more efficient TMRCA computation?

### DISCUSSION THREAD

**Discussion by feilchenfeldt** [tskit-dev/tskit#1306] [Ideas]
Hello,

I recently started to use tskit in python to iterate over tree sequences and interrogate the trees.
I wrote a function for getting the distance in the tree between two samples. Now I realised that tree.tmrca() is implemented, which basically gives the same thing.

However, I see that my function for getting the tmrca is consistently about 10 times faster than tree.tmrca() on my trees of 486 samples, while giving identical output.

I was trying to understand the implemented method, but the actuall code is burried in `_ll` objects which are obscure to me.

I post my code below and wanted to ask whether there is anything that speaks against implementing .tmrca that way, given that it seems to be much faster.

Analogously to TMRCA, one could also retrieve the id of the MRCA analogously (just returning i).

Thanks,
Hannes

```
def get_tmrca(tree, i, j):
    t0 = tree.get_time(i)
    t1 = tree.get_time(j)
    while i != j:
        ti = tree.get_time(i)
        tj = tree.get_time(j)
        if ti < tj:
            i = tree.get_parent(i)
        else:
            j = tree.get_parent(j)
    tmrca = tree.get_time(i) - (t0 - t1)/2
    return tmrca
```

**Comment by petrelharp** [comment]
Hm, interesting. That's a very nice and clear function! The `.tmrca( )` method is implemented in C ([here's the algorithm](https://github.com/tskit-dev/tskit/blob/main/c/tskit/trees.c#L3490)), which is why it's obscured. I'm surprised that it's faster for you? I did a quick test on a much bigger tree that shows the C method scaling much better:
```
t = msprime.simulate(1000000).first()
# almost instant
sum([t.tmrca(0,j) for j in range(1000000)])
# takes quite a long time
sum([get_tmrca(t,0,j) for j in range(1000000)])
```
This isn't a very careful timing test, of course.

In your case, was the built-in method actually taking a long time?

**Reply by feilchenfeldt** [reply]
Dear Peter, 

thanks for your reply and for testing. That is indeed curious. I can replicate what you are saying about trees produced by msprime, and even when running msprime on the same amount of samples as in my dataset (486), I see that the C-implemented tree.get_tmrca() is around 20 times faster than my python function.

However, this is very different for my trees that are based on actual data. I produced them by running RELATE on statistically phased variants from resequencing of 243 cichlid fishes and let RELATE output tree sequences. On this dataset the python function is 10x faster than the C method for the first tree in the tree sequence and around 2 times faster than the C method when run on random pairs of samples for each of the ~40000 trees in the tree sequence of one chromosome.

I uploaded the output of `tree.as_dict_of_dicts()` for the first tree as a `.json` file here:

https://drive.google.com/file/d/1hpe7sp_XKQJTl2NkfvgYwi9mjJ1Z5EC5/view?usp=sharing

If you have any idea why the C method performs relatively so much worse on this tree, I would be very curious to know. I can also share the tree in any other format if you let me know.

Here the timing comparsions for this tree:

C method
```
%%timeit
for node1, node2 in itertools.combinations(tree.samples(), 2):
    tree.get_tmrca(node1, node2)
1 loop, best of 3: 25.9 s per loop
```

python function
```
%%timeit
for node1, node2 in itertools.combinations(tree.samples(), 2):
    get_tmrca(tree, node1, node2)
1 loop, best of 3: 2.82 s per loop
```

**Reply by jeromekelleher** [reply]
Aha, thanks for the updates @feilchenfeldt! I think I see the issue here: in the tskit version we're allocating two blocks of memory to store the list of nodes up to root. To keep things simple, we allocate ``ts.num_nodes`` slots in these arrays. This is fine for msprime produced tree sequences, because we typically have around O(n + T log(n)) nodes, there T is the number of trees. For RELATE though, we have O(nT) nodes, because it basically stores each tree independently. So, for RELATE trees, we're allocating a huge block of memory for each get_mrca computation, which apparently is taking quite a lot of time.

We should change tskit's C code to use your algorithm, it's *much* nicer. Would you be interested in helping out with this (contributions count towards authorship of the eventual tskit 1.0 paper)?

**Reply by feilchenfeldt** [reply]
Dear Jerome,

thanks for your reply. I would be interested, however, my last time actively using C was 15 years ago. But I have a student who might be better at that. I could talk to her. First we will need to understand what the C code exactly does as to not break anything. 
How to we best go about this?

Thanks,
Hannes

**Reply by jeromekelleher** [reply]
The C code is pretty well self-contained in [this function](https://github.com/tskit-dev/tskit/blob/main/c/tskit/trees.c#L3490).  The directions for getting started are in the [developer docs](https://tskit.dev/tskit/docs/stable/development.html#c-library). If the student is interested we can invite her to the tskit-dev Slack workspace to help her get up to speed.

I've created #1309 to track.

**Reply by jeromekelleher** [reply]
Hi @feilchenfeldt - just wondering if you're interested in following up on this? I've been looking into this and I can make the change pretty easily, but if someone else was interested in doing it I'd be happy to help them through the process.

**Reply by feilchenfeldt** [reply]
Hello @jeromekelleher,

We would like to contribute to tskit in the medium term. 
There are some other features/improvements which we could not find and would like to see implemented. 
(For example, a quick way of subsetting a tree to a subset of the samples and their histories and get the topolgy for these samples. We found that functionality for a tree sequence, but not for a single tree.  Is the discussion forum the right place to ask about such things?) 

For the current problem at hand, if you have looked into this I am sure you could implement it much faster and better, so please go ahead. Would that also improve the low efficiency with relate trees, or is this something the relate developpers should address? (Should we contact them?)

Best wishes,
Hannes

I actually think the algorithm above could be slightly more efficient like this:

```
def get_mrca(tree, i, j):
    t0 = tree.get_time(i)
    t1 = tree.get_time(j)
    ti = t0
    tj = t1
    while i != j:
        if ti < tj:
            i = tree.get_parent(i)
            ti = tree.get_time(i)
        else:
            j = tree.get_parent(j)
            tj = tree.get_time(j)
    return i

```

**Reply by jeromekelleher** [reply]
> We would like to contribute to tskit in the medium term.

Excellent!

> We found that functionality for a tree sequence, but not for a single tree. Is the discussion forum the right place to ask about such things?

The discussion forum is the perfect place for these discussions. Please go ahead and start as many threads as you like to explore the functionality you'd like to see.

> Would that also improve the low efficiency with relate trees, or is this something the relate developpers should address? (Should we contact them?)

I'm not sure there much that can be done for RELATE trees, tbh - it's a fundamental property of the algorithm that node identities aren't shared between trees, and that's what tskit relies on efficiency. There might be some improvements that could be made, though, so there's no harm in letting Leo know the issues you're hitting.

> For the current problem at hand, if you have looked into this I am sure you could implement it much faster and better, so please go ahead. 

OK I'll implement the tmrca and ping you for review.

**Reply by feilchenfeldt** [reply]
Great, my PhD student Jaysmita @JSaha20 will be also interested in understanding the changes and reviewing.

**Reply by benjeffery** [reply]
Here's the PR that resulted from this for the benefit of future readers: https://github.com/tskit-dev/tskit/pull/1313
